Update the API "login" action documentation (#286)
kuzzleio/documentation: src/protocols/1/native-protocols/mqtt/index.md

---
layout: full.html.hbs
title: MQTT
order: 0
---
# MQTT

<div class="alert alert-warning">
By default, the MQTT protocol is disabled in Kuzzle configuration.
</div

---

## Configuration

> The protocol can be configured via [Kuzzle'rc configuration]({{ site_base_path }}guide/1/essentials/configuration/), under ``server > protocols > mqtt`` section.

Author: ```suggestion
The protocol can be configured via the [kuzzlerc configuration file]({{ site_base_path }}guide/1/essentials/configuration/), under the ``server > protocols > mqtt`` section.
```

| Option | Type | Description | Default |
|---|---|---|---|

> | ``enabled`` | <pre>boolean</pre> | Enables/Disables the MQTT protocol support | ``false`` |

Author: ```suggestion
| ``enabled`` | <pre>boolean</pre> | Enable/Disable the MQTT protocol support | ``false`` |
```

| ``allowPubSub`` | <pre>boolean</pre> | Allow MQTT pub/sub capabilities or restrict to Kuzzle requests only | ``false`` |
| ``developmentMode`` | <pre>boolean</pre> | Switches ``responseTopic`` back to a regular public topic | ``false`` |
| ``disconnectDelay`` | <pre>integer</pre> | Delay in ms to apply between a disconnection notification is received and the connection is actually removed | 250 |
| ``requestTopic`` | <pre>string</pre> | Name of the topic listened by the plugin for requests | ``"Kuzzle/request"`` |
| ``responseTopic`` | <pre>string</pre> | Name of the topic clients should listen to get requests result | ``"Kuzzle/response"`` |
| ``server`` | <pre>object</pre> | Constructor options pased to underlying mqtt server. See [mosca documentation](https://github.com/mcollina/mosca/wiki/Mosca-advanced-usage#other-options-of-mosca-the-ones-we-inserted-in-our-moscasettings-var) for further reference. | ``{port: 1883}`` |

example:

``.kuzzlerc``
```json
{
  "server": {
    "protocols": {
      "mqtt": {
        "allowPubSub": true
      }
    }
  }
}
```

## How to use

### Sending an API request and getting the response

> By default, the MQTT protocol listens to the ``Kuzzle/request`` MQTT topic (see [configuration](#configuration)) for requests to the [Kuzzle API](https://docs.kuzzle.io/api/1/essentials/connecting-to-kuzzle/).

Author: ```suggestion
By default, the MQTT protocol listens to the ``Kuzzle/request`` MQTT topic (see [configuration](#configuration)) for requests to the [Kuzzle API]({{ site_base_path }}api/1/essentials/connecting-to-kuzzle/).
```

    if (message.requestId === 'some unique ID') {
      channels.push(message.result.channel);
      client.subscribe(message.result.channel);
    }
  }
  else if (channels.indexOf(topic) !== -1) {
    // Subscription notification
    console.log('Notification: ', message);
  }
});
```

## Authorizations

### Publishing

If ``allowPubSub`` is set to ``false``, clients can only publish to the ``requestTopic`` topic (defaults to ``Kuzzle/request``).

If ``allowPubSub`` is set to ``true``, clients are only forbidden to publish to the ``responseTopic`` topic (defaults to ``Kuzzle/response``).

<div class="alert alert-warning">
    Wildcards subcriptions are not allowed
</div

If a client tries to publich to an unauthorized topic, his connection will immediately be shut down by the server.

### Subscribing

Subscription attempts to the ``requestTopic`` topic (defaults to ``Kuzzle/request``) are ignored: client requests can only be listened by the MQTT server.

## Development mode

The MQTT ``Kuzzle/response`` topic is by default a special topic that acts as a private channel. Each client receives its own responses only, offering a simple first security layer.

> While this behaviour is urgently recommended in production, it can bring a small drawback when testing and developing applications: it does not allow using most CLI tools.  

Author: ```suggestion
While this behavior is urgently recommended in production, it can bring a small drawback when testing and developing applications: it does not allow using most CLI tools.  
```

> Many CLI tools, such as Mosquitto offter two separate binaries, one for subscribing and one for publishing. These act as two different clients and the subscriber won't receive any response sent to the publisher by default.

Author: ```suggestion
Many CLI tools, such as Mosquitto offer two separate binaries, one for subscribing and one for publishing. These act as two different clients and the subscriber won't receive any response sent to the publisher by default.
```
